src/log.rs

```diff
@@ -81,13 +82,316 @@ type CompatFilterStream<T, R> = Compat01As03<FilterStream<T, R>>;
+        let web3 = builder.web3.clone();
+        let to_block = builder.to_block.unwrap_or(BlockNumber::Latest);
+        let block_page_size = builder.block_page_size.unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+
+        let page_block = match builder.from_block.unwrap_or(BlockNumber::Latest) {
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+        let end_block = match to_block {
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+
+        let filter = builder.into_filter();
+
+        PastLogsPager {
+            web3,
+            to_block,
+            block_page_size,
+            filter,
+            page_block,
+            end_block,
+        }
+    }
+
+    async fn next_page(&mut self) -> Result<Option<Vec<Log>>, ExecutionError> {
+        let mut page_block = match self.page_block {
+            Some(value) => value,
+            None => {
+                // NOTE: This means the `from_block` is `Latest` or `Pending, so
+                //   just do a single query.
+                let page = self
+                    .web3
+                    .eth()
+                    .logs(self.filter.clone().build())
+                    .compat()
+                    .await?;
+
+                // Ensure that subsequent calls to `next_page` return `None`.
+                self.page_block = Some(1);
+                self.end_block = Some(0);
+
+                if page.is_empty() {
+                    return Ok(None);
+                } else {
+                    return Ok(Some(page));
+                }
+            }
+        };
+
+        let end_block = match self.end_block {
+            Some(value) => value,
+            None => {
+                let latest_block = self.web3.eth().block_number().compat().await?.as_u64();
+                self.end_block = Some(latest_block);
+                latest_block
+            }
+        };
+
+        while page_block <= end_block {
+            // NOTE: Log block ranges are inclusive.
+            let page_end = page_block + self.block_page_size - 1;
+            let page_to_block = if page_end < end_block {
+                BlockNumber::Number(page_end.into())
+            } else {
+                // NOTE: The last page is handled a bit differently
+                //   by using the `to_block` that was originally
+                //   specified to the builder. This is done in case
+                //   the to block was "latest" or "pending", where
+                //   we want to make sure that the last call
+                //   includes blocks that have been added since the
+                //   start of the paginated stream.
+                self.to_block
+            };
```

Comment:
It would also be sensible to say that the last block is determined when PastLogsParser is created which could make the implementation here simpler in several places.